receiver/splunkenterprisereceiver/scraper.go

```diff
@@ -2,3 +2,196 @@
 // SPDX-License-Identifier: Apache-2.0
 
 package splunkenterprisereceiver // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/splunkenterprisereceiver"
+
+import (
+	"context"
+	"encoding/json"
+	"encoding/xml"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"strconv"
+	"time"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+	"go.opentelemetry.io/collector/receiver"
+	"go.opentelemetry.io/collector/receiver/scrapererror"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/splunkenterprisereceiver/internal/metadata"
+)
+
+var (
+	errMaxSearchWaitTimeExceeded = errors.New("Maximum search wait time exceeded for metric")
+)
+
+type splunkScraper struct {
+	splunkClient *splunkEntClient
+	settings     component.TelemetrySettings
+	conf         *Config
+	mb           *metadata.MetricsBuilder
+}
+
+func newSplunkMetricsScraper(params receiver.CreateSettings, cfg *Config) splunkScraper {
+	return splunkScraper{
+		settings: params.TelemetrySettings,
+		conf:     cfg,
+		mb:       metadata.NewMetricsBuilder(cfg.MetricsBuilderConfig, params),
+	}
+}
+
+// Create a client instance and add to the splunkScraper
+func (s *splunkScraper) start(_ context.Context, _ component.Host) (err error) {
+	c := newSplunkEntClient(s.conf)
+	s.splunkClient = &c
+	return nil
+}
+
+// The big one: Describes how all scraping tasks should be performed. Part of the scraper interface
+func (s *splunkScraper) scrape(ctx context.Context) (pmetric.Metrics, error) {
+	errs := &scrapererror.ScrapeErrors{}
+	now := pcommon.NewTimestampFromTime(time.Now())
+
+	s.scrapeLicenseUsageByIndex(ctx, now, errs)
+	s.scrapeIndexThroughput(ctx, now, errs)
+	return s.mb.Emit(), errs.Combine()
+}
+
+// Each metric has its own scrape function associated with it
+func (s *splunkScraper) scrapeLicenseUsageByIndex(ctx context.Context, now pcommon.Timestamp, errs *scrapererror.ScrapeErrors) {
+	var sr searchResponse
+	// Because we have to utilize network resources for each KPI we should check that each metrics
+	// is enabled before proceeding
+	if !s.conf.MetricsBuilderConfig.Metrics.SplunkLicenseIndexUsage.Enabled {
+		return
+	}
+
+	sr = searchResponse{
+		search: searchDict[`SplunkLicenseIndexUsageSearch`],
+	}
+
+	start := time.Now()
+	req, err := s.splunkClient.createRequest(ctx, &sr)
+	if err != nil {
+		errs.Add(err)
+	}
+
+	res, err := s.splunkClient.makeRequest(req)
+	if err != nil {
+		errs.Add(err)
+	}
+	defer res.Body.Close()
+
+	err = unmarshallSearchReq(res, &sr)
+	if err != nil {
+		errs.Add(err)
+	}
+
+	for ok := true; ok; ok = (sr.Return == 204) {
+		req, err = s.splunkClient.createRequest(ctx, &sr)
+		if err != nil {
+			errs.Add(err)
+		}
+
+		res, err = s.splunkClient.makeRequest(req)
+		if err != nil {
+			errs.Add(err)
+		}
+		defer res.Body.Close()
```

Comment:
that defer will run when the function exits, so you might be piling on unclosed responses. This means that the client cannot reuse the same http connection when it connects to Splunk and that means additional TCP connections to the server. This might be your intent, but if it isn't, I'd recommend closing the body after unmarshalling it if no errors occur.